app/src/cose.h

```diff
@@ -27,30 +27,161 @@
+  // profile header parameter yet, but it's useful to have a label so we can
+  // easily act on the absence of a profile.
+  static constexpr int64_t COSE_HEADER_PARAM_PROFILE = 6861;
+
   // Temporary assignments from draft-birkholz-scitt-architecture
   static constexpr int64_t COSE_HEADER_PARAM_ISSUER = 391;
   static constexpr int64_t COSE_HEADER_PARAM_FEED = 392;
   static constexpr int64_t COSE_HEADER_PARAM_SCITT_RECEIPTS = 394;
 
+  // Notary header parameters.
+  static constexpr const char* NOTARY_HEADER_PARAM_SIGNING_SCHEME =
+    "io.cncf.notary.signingScheme";
+  static constexpr const char* NOTARY_HEADER_PARAM_SIGNING_TIME =
+    "io.cncf.notary.signingTime";
+  static constexpr const char* NOTARY_HEADER_PARAM_AUTHENTIC_SIGNING_TIME =
+    "io.cncf.notary.authenticSigningTime";
+  static constexpr const char* NOTARY_HEADER_PARAM_EXPIRY =
+    "io.cncf.notary.expiry";
+
   struct COSEDecodeError : public std::runtime_error
   {
     COSEDecodeError(const std::string& msg) : std::runtime_error(msg) {}
   };
 
+  struct UnprotectedHeader
+  {
+    // We currently expect only notary to use the unprotected header and
+    // we expect to find only the x5chain in there.
+    std::optional<std::vector<std::vector<uint8_t>>> x5chain;
+  };
+
+  UnprotectedHeader decode_unprotected_header(
+    const std::vector<uint8_t>& cose_sign1)
+  {
+    UnprotectedHeader parsed;
+    // Adapted from parse_cose_header_parameters in t_cose_parameters.c.
+    // t_cose doesn't support custom header parameters yet.
+
+    QCBORError qcbor_result;
+
+    QCBORDecodeContext ctx;
+    QCBORDecode_Init(
+      &ctx, cbor::from_bytes(cose_sign1), QCBOR_DECODE_MODE_NORMAL);
+    QCBORDecode_EnterArray(&ctx, nullptr);
+    qcbor_result = QCBORDecode_GetError(&ctx);
+    if (qcbor_result != QCBOR_SUCCESS)
+    {
+      throw COSEDecodeError("Failed to parse COSE_Sign1 outer array");
+    }
+
+    uint64_t tag = QCBORDecode_GetNthTagOfLast(&ctx, 0);
+    if (tag != CBOR_TAG_COSE_SIGN1)
+    {
+      throw COSEDecodeError("COSE_Sign1 is not tagged");
+    }
+    QCBORItem coseItem;
+    QCBORDecode_GetNext(&ctx, &coseItem);
+
+    struct q_useful_buf_c unprotected_parameters;
+    QCBORDecode_EnterMap(&ctx, NULL);
+
+    enum
+    {
+      ALG_INDEX,
+      X5CHAIN_INDEX,
+      END_INDEX,
+    };
+    QCBORItem header_items[END_INDEX + 1];
+    header_items[X5CHAIN_INDEX].label.int64 = COSE_HEADER_PARAM_X5CHAIN;
+    header_items[X5CHAIN_INDEX].uLabelType = QCBOR_TYPE_INT64;
+    header_items[X5CHAIN_INDEX].uDataType = QCBOR_TYPE_ANY;
+    header_items[END_INDEX].uLabelType = QCBOR_TYPE_NONE;
+
+    QCBORDecode_GetItemsInMap(&ctx, header_items);
+
+    qcbor_result = QCBORDecode_GetError(&ctx);
+    if (qcbor_result != QCBOR_SUCCESS)
+    {
+      throw COSEDecodeError("Failed to decode unprotected header");
+    }
+    if (header_items[X5CHAIN_INDEX].uDataType != QCBOR_TYPE_NONE)
+    {
+      parsed.x5chain = std::vector<std::vector<uint8_t>>();
```

Comment:
This code already exists in `decode_protected_header`. It should be refactored into a common function (it's fine if the exception messages don't mention protected/unprotected).